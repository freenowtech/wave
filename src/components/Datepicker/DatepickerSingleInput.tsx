import { FirstDayOfWeek, START_DATE } from '@datepicker-react/hooks';
import parse from 'date-fns/parse';
import React, { ChangeEventHandler, FC, MutableRefObject, useEffect, useRef, useState } from 'react';
import TetherComponent from 'react-tether';
import { MarginProps, WidthProps } from 'styled-system';
import { Input } from '../Input/Input';

import { Datepicker } from './Datepicker';
import { isValidDateText } from './utils/isValidDateText';
import { Elevation } from '../../essentials';
import { useGeneratedId } from '../../utils/hooks/useGeneratedId';
import { HelperText } from '../HelperText/HelperText';
import { dateToDisplayText } from './utils/dateToDisplayText';
import { useLocaleObject } from './utils/useLocaleObject';

interface DatepickerSingleInputProps extends MarginProps, WidthProps {
    /**
     * Placeholder for the input.
     */
    placeholder?: string;
    /**
     * Label for the input.
     */
    label?: string;
    /**
     * Function that used when datepicker closes without selected date.
     */
    onClose?: () => void;
    /**
     * Function that used when datepicker selects new date.
     */
    onChange?: (date?: Date) => void;
    /**
     * Minimal date to select from.
     */
    minDate?: Date;
    /**
     * Maximum date to select from.
     */
    maxDate?: Date;
    /**
     * Accepts a number for first day of the week from 0 (Sunday) to 6 (Saturday).
     * Default: 1
     */
    firstDayOfWeek?: FirstDayOfWeek;
    /**
     * Function that runs for each date and returns boolean whether date is disabled or not.
     * Default: () => false
     */
    isDateBlocked?: (date: Date) => boolean;
    /**
     * String to format dates.
     * Default: dd/MM/yyyy
     */
    displayFormat?: string;
    /**
     * String to define the locale in ISO-639-1
     * Default: en-US
     */
    locale?: string;
    /**
     * Set the value of the input
     */
    value?: Date;
    /**
     * Text to be shown if error filling the input
     * or fn to be trigger as a callback when error
     */
    errorHandler?: (() => void) | string;
    /**
     * The id to be assigned to the input field
     */
    inputId?: string;
}

const DatepickerSingleInput: FC<DatepickerSingleInputProps> = ({
    minDate,
    maxDate,
    firstDayOfWeek,
    isDateBlocked,
    onClose,
    onChange,
    placeholder,
    label,
    displayFormat = 'dd/MM/yyyy',
    locale = 'en-US',
    value,
    errorHandler,
    inputId,
    ...rest
}: DatepickerSingleInputProps) => {
    const localeObject = useLocaleObject(locale);
    const inputRef = useRef<HTMLDivElement>();
    const [isFocused, setIsFocused] = useState(false);
    const [inputText, setInputText] = useState(dateToDisplayText(localeObject, displayFormat, value));
    const [error, setError] = useState(false);
    const displayErrorMessage = typeof errorHandler === 'string';

    const id = useGeneratedId(inputId);

    useEffect(() => {
        if (error && typeof errorHandler === 'function') {
            errorHandler();
        }
    }, [error]);

    useEffect(() => {
        setInputText(dateToDisplayText(localeObject, displayFormat, value));
    }, [value, localeObject, displayFormat]);

    useEffect(() => {
        // after opening the datepicker, bring back focus to the input
        const target = inputRef.current && (inputRef.current.children[0] as HTMLInputElement);

        if (target && isFocused) {
            target.focus();
        }
    }, [isFocused]);

    const handleDatepickerClose = () => {
        setIsFocused(false);

        if (onClose) {
            onClose();
        }
    };

    const handleDateChange = (date?: Date) => {
        if (onChange) {
            onChange(date);
        }
    };

    const handleDateTextChange: ChangeEventHandler<HTMLInputElement> = ({ target: { value: inputValue } }) => {
        setInputText(inputValue);

        const parsedDate = inputValue === '' ? undefined : parse(inputValue, displayFormat, new Date());
        const validDate = isValidDateText(parsedDate, inputValue, displayFormat);
        const hasError = !!inputValue && !validDate;

        if (!hasError) {
            handleDateChange(parsedDate);
        }

        setError(hasError);
    };

    return (
        <TetherComponent
            attachment="top left"
            targetAttachment="bottom left"
            constraints={[
                {
                    to: 'window',
                    attachment: 'together'
                }
            ]}
            renderTarget={(ref: MutableRefObject<HTMLDivElement>) => (
                <>
                    <Input
                        ref={element => {
                            inputRef.current = element;
                            // eslint-disable-next-line no-param-reassign
                            ref.current = element;
                        }}
                        id={id}
                        autoComplete="off"
                        className="startDate"
                        data-testid="start-date-input"
                        label={label}
                        placeholder={placeholder}
                        value={inputText}
                        onFocus={() => setIsFocused(true)}
                        onBlur={handleDatepickerClose}
                        onChange={handleDateTextChange}
                        data-error={error}
                        {...rest}
                    />
                    {displayErrorMessage && error && !isFocused && (
                        <HelperText mt="1">{errorHandler || `error (${displayFormat})`}</HelperText>
                    )}
                </>
            )}
            renderElement={(ref: MutableRefObject<HTMLDivElement>) =>
                isFocused && (
                    <Datepicker
                        ref={ref}
                        numberOfMonths={1}
                        exactMinBookingDays
                        minBookingDays={1}
                        startDate={value}
                        endDate={value}
                        minBookingDate={minDate}
                        maxBookingDate={maxDate}
                        firstDayOfWeek={firstDayOfWeek}
                        focusedInput={isFocused ? START_DATE : undefined}
                        onDatesChange={({ focusedInput, startDate }) => {
                            setIsFocused(focusedInput !== null);
                            handleDateChange(startDate || undefined);
                        }}
                        isDateBlocked={isDateBlocked}
                        locale={localeObject}
                    />
                )
            }
            style={{ zIndex: Elevation.DATEPICKER }}
        />
    );
};

export { DatepickerSingleInput, DatepickerSingleInputProps };
