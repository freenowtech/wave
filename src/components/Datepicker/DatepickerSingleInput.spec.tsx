import { fireEvent, render, waitFor, screen } from '@testing-library/react';
import { format } from 'date-fns';
import userEvent from '@testing-library/user-event';
import { advanceTo, clear } from 'jest-date-mock';
import * as React from 'react';
import { DatepickerSingleInput } from './DatepickerSingleInput';

describe('DatepickerSingleInput', () => {
    const today = new Date(1999, 1, 2);

    beforeAll(() => {
        advanceTo(today);
    });

    afterAll(() => {
        clear();
    });

    it('renders the default props', () => {
        expect(render(<DatepickerSingleInput />).container).toMatchSnapshot();
    });

    it('can be disabled', () => {
        const { getByRole } = render(<DatepickerSingleInput disabled />);
        const input = getByRole('textbox');
        expect(input).toBeDisabled();
    });

    describe('should call onClose function', () => {
        it('when clicking outside', async () => {
            const mockCloseHandler = jest.fn();
            const { getByTestId } = render(<DatepickerSingleInput onClose={mockCloseHandler} />);

            // Open datepicker
            userEvent.click(getByTestId('start-date-input'));

            // Click outside
            fireEvent.blur(getByTestId('start-date-input'));

            await waitFor(() => expect(mockCloseHandler).toHaveBeenCalled());
        });
    });

    it('should call onChange function after the date selection', async () => {
        const mockChangeHandler = jest.fn();
        const { getByTestId, getByText } = render(<DatepickerSingleInput onChange={mockChangeHandler} />);

        // Open datepicker
        userEvent.click(getByTestId('start-date-input'));

        // Select the first date
        fireEvent.click(getByText('01'));

        const currentMonth = today.getMonth();
        const currentYear = today.getFullYear();

        await waitFor(() => expect(mockChangeHandler).toHaveBeenCalledWith(new Date(currentYear, currentMonth, 1)));
    });

    it('calls errorHandler as a fn if a callback is provided', () => {
        const mockChangeHandler = jest.fn();
        const mockErrorTextHandler = jest.fn();
        const { getByTestId } = render(
            <DatepickerSingleInput onChange={mockChangeHandler} errorHandler={mockErrorTextHandler} />
        );
        const input = getByTestId('start-date-input');

        const intendedDate = new Date(2020, 5, 1);
        // force to use WRONG FORMAT to test the error
        const formattedDate = format(intendedDate, 'dd/MM/yy');

        // Simulate an input change
        userEvent.type(input, formattedDate);

        expect(mockErrorTextHandler).toHaveBeenCalledTimes(1);
    });

    it('calls onChange function after update input value', () => {
        const mockChangeHandler = jest.fn();
        const { getByTestId } = render(<DatepickerSingleInput onChange={mockChangeHandler} />);
        const input = getByTestId('start-date-input');

        // Focus on datepicker
        userEvent.click(input);

        const currentMonth = today.getMonth();
        const currentYear = today.getFullYear();
        const intendedDate = new Date(currentYear, currentMonth, 1);
        const formattedDate = format(intendedDate, 'dd/MM/yyyy');

        // Simulate an input change
        userEvent.type(input, formattedDate);

        expect(mockChangeHandler).toHaveBeenCalledWith(intendedDate);
    });

    it('renders an error feedback if gets filled with invalid value', () => {
        const mockChangeHandler = jest.fn();
        const { getByTestId } = render(<DatepickerSingleInput onChange={mockChangeHandler} />);
        const input = getByTestId('start-date-input');

        // Focus on datepicker
        userEvent.click(input);

        const currentMonth = today.getMonth();
        const currentYear = today.getFullYear();
        const intendedDate = new Date(currentYear, currentMonth, 1);
        // force to use WRONG FORMAT to test the error
        const formattedDate = format(intendedDate, 'dd/MM/yy');

        // Simulate an input change
        userEvent.type(input, formattedDate);

        expect(mockChangeHandler).not.toHaveBeenCalled();
        expect(input).toHaveAttribute('data-error', 'true');

        // Simulate an input empty again so error state should disappear
        fireEvent.change(input, { target: { value: '' } });
        expect(input).toHaveAttribute('data-error', 'false');
    });

    it('closes the calendar of the first datepicker and opens the calendar of the second', () => {
        const { getAllByTestId, getAllByText } = render(
            <>
                <DatepickerSingleInput />
                <DatepickerSingleInput />
            </>
        );

        const inputs = getAllByTestId('start-date-input');
        expect(inputs).toHaveLength(2);

        userEvent.click(inputs[0]);
        expect(getAllByText('01')).toHaveLength(1);

        userEvent.click(inputs[1]);
        expect(getAllByText('01')).toHaveLength(1);
    });

    it('localizes selected date correctly', async () => {
        const date = new Date(2020, 11, 15);
        render(<DatepickerSingleInput value={date} locale="de-DE" displayFormat="dd MMMM yyyy" />);

        const localizedCurrentMonth = 'Dezember';
        const currentYear = date.getFullYear();

        expect(await screen.findByDisplayValue(`15 ${localizedCurrentMonth} ${currentYear}`)).toBeInTheDocument();
    });

    it('localizes datepicker calendar month and weekdays', async () => {
        const date = new Date(2020, 11, 15);
        render(<DatepickerSingleInput value={date} locale="de-DE" />);

        // Open datepicker
        userEvent.click(screen.getByTestId('start-date-input'));

        const localizedCurrentMonth = 'Dezember';
        const currentYear = date.getFullYear();

        expect(await screen.findByText(`${localizedCurrentMonth} ${currentYear}`)).toBeInTheDocument();
        expect(await screen.findByText('Do')).toBeInTheDocument();
    });
});
