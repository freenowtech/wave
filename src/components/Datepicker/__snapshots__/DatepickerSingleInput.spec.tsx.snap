// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatepickerSingleInput renders the default props 1`] = `
.c1 {
  margin: 0;
  box-sizing: border-box;
  background: #FFFFFF;
  border-radius: 0;
  color: #001E3E;
  font-size: 1rem;
  font-family: "Open Sans",sans-serif;
  -webkit-transition: box-shadow 100ms,border 100ms;
  transition: box-shadow 100ms,border 100ms;
  outline: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 100%;
  border-radius: 0.25rem;
  border: 0.0625rem solid #C6CDD4;
  font-size: 1rem;
  height: 3rem;
  padding: 0 0.75rem;
}

.c1::-webkit-input-placeholder {
  color: #9CA7B4;
}

.c1::-moz-placeholder {
  color: #9CA7B4;
}

.c1:-ms-input-placeholder {
  color: #9CA7B4;
}

.c1::placeholder {
  color: #9CA7B4;
}

.c1:active,
.c1:focus {
  border-color: #096BDB;
  box-shadow: inset 0 0 0 0.0625rem #096BDB;
}

.c1:disabled {
  color: #C6CDD4;
  border-color: #C6CDD4;
  box-shadow: none;
  cursor: not-allowed;
}

.c1:disabled::-webkit-input-placeholder {
  color: #C6CDD4;
}

.c1:disabled::-moz-placeholder {
  color: #C6CDD4;
}

.c1:disabled:-ms-input-placeholder {
  color: #C6CDD4;
}

.c1:disabled::placeholder {
  color: #C6CDD4;
}

.c1:-webkit-autofill,
.c1:-webkit-autofill:hover,
.c1:-webkit-autofill:focus,
.c1:-webkit-autofill:active {
  -webkit-text-fill-color: #001E3E;
  -webkit-transition: background-color 99999999ms ease 99999999ms;
  transition: background-color 99999999ms ease 99999999ms;
}

.c1 + .sc-AxhUy {
  color: #9CA7B4;
  background: #FFFFFF;
  background: linear-gradient(0deg,#FFFFFF calc(50% + 0.0625rem),transparent 50%);
}

.c1:disabled + .sc-AxhUy {
  color: #C6CDD4;
}

.c1:-webkit-autofill + .sc-AxhUy,
.c1:-webkit-autofill:hover + .sc-AxhUy,
.c1:-webkit-autofill:focus + .sc-AxhUy,
.c1:-webkit-autofill:active + .sc-AxhUy {
  font-weight: 600;
  top: -0.5rem;
  font-size: 0.75rem;
}

.c1:focus:not(:disabled) + .sc-AxhUy {
  font-weight: 600;
  top: -0.5rem;
  font-size: 0.75rem;
  color: #096BDB;
  background: #FFFFFF;
  background: linear-gradient(0deg,#FFFFFF calc(50% + 0.0625rem),transparent 50%);
}

.c0 {
  display: inline-block;
  position: relative;
  box-sizing: border-box;
}

<div>
  <div
    class="c0 startDate"
  >
    <input
      autocomplete="off"
      class="sc-AxjAm c1"
      data-error="false"
      data-testid="start-date-input"
      id="random"
      type="text"
      value=""
    />
  </div>
</div>
`;
